gapis/resolve/dependencygraph2/dependency_graph.go

```diff
@@ -0,0 +1,425 @@
+		cmdID = api.CmdID(i) | initCmdBit
+		if err := cb(subctx, cmdID, cmd); err != nil {
+			if err != api.Break {
+				return err
+			}
+			return nil
+		}
+		if err := task.StopReason(ctx); err != nil {
+			return err
+		}
+	}
+
+	for i, cmd := range g.capture.Commands {
+		cmdID = api.CmdID(i)
+		if err := cb(subctx, cmdID, cmd); err != nil {
+			if err != api.Break {
+				return err
+			}
+			return nil
+		}
+		if err := task.StopReason(ctx); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// ForeachNode iterates over all nodes in the graph
+func (g *dependencyGraph) ForeachNode(cb func(NodeID, Node) error) error {
+	for i, node := range g.nodes {
+		err := cb(NodeID(i), node)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// ForeachDependency iterates over all pairs (src, tgt), where src depends on tgt
+func (g *dependencyGraph) ForeachDependency(cb func(NodeID, NodeID) error) error {
+	for i, depsFrom := range g.dependenciesFrom {
+		src := NodeID(i)
+		for _, tgt := range depsFrom {
+			err := cb(src, tgt)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// ForeachDependencyFrom iterates over all the nodes tgt, where src depends on tgt
+func (g *dependencyGraph) ForeachDependencyFrom(src NodeID, cb func(NodeID) error) error {
+	for _, tgt := range g.dependenciesFrom[src] {
+		err := cb(tgt)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// ForeachDependencyTo iterates over all the nodes src, where src depends on tgt.
+// If Config().ReverseDependencies is false, this will return an error.
+func (g *dependencyGraph) ForeachDependencyTo(tgt NodeID, cb func(NodeID) error) error {
+	for _, src := range g.dependenciesTo[tgt] {
+		err := cb(src)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Capture returns the capture whose dependencies are stored in this graph
+func (g *dependencyGraph) Capture() *capture.Capture {
+	return g.capture
+}
+
+// GetCommand returns the command identified by the given CmdID
+func (g *dependencyGraph) GetCommand(cmdID api.CmdID) api.Cmd {
+	if cmdID.IsReal() {
+		if cmdID >= api.CmdID(len(g.capture.Commands)) {
+			return nil
+		}
+		return g.capture.Commands[cmdID]
+	} else {
+		cmdID &= ^initCmdBit
+		if cmdID >= api.CmdID(len(g.initialCommands)) {
+			return nil
+		}
+		return g.initialCommands[cmdID]
+	}
+}
+
+// InitialCommands returns the initial commands, which
+// reconstruct the initial state before the first command in the capture.
+func (g *dependencyGraph) InitialCommands() []api.Cmd {
+	return g.initialCommands
+}
+
+// InitialCmdRanges returns the memory ranges used by the initial commands.
+func (g *dependencyGraph) InitialCmdRanges() interval.U64RangeList {
+	return g.initialCmdRanges
+}
+
+// Config returns the config used to create this graph
+func (g *dependencyGraph) Config() DependencyGraphConfig {
+	return g.config
+}
+
+func (g *dependencyGraph) setDependencies(src NodeID, targets []NodeID) {
```

Comment:
Maybe we can have `addDependencies(src NodeID, tgts ...NodeID)` instead of two functions `setDependencies()` and `addDependency`

Reply:
`setDependencies` exists for performance.

Using `addDependency` instead would result in `dependenciesFrom[src]` to be dynamically grown by `append`. This would be slow, and also result in `dependenciesFrom[src]` being significantly larger than necessary, on average. 

Instead, the `DependencyGraphBuilder` creates a slice with almost* exactly the right capacity, and assigns dependenciesFrom directly, using `setDependencies`. (*forward dependencies make this slightly more complicated, but are much more rare than backward dependencies).